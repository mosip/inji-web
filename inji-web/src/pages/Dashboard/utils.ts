export const convertStringIntoPascalCase = (text: string | null) => {
    if (text != null) {
        return text
            .toLocaleLowerCase()
            .split(" ")
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(" ");
    }
};
