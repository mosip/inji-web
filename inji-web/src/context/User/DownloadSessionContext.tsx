import {CredentialTypeDisplayArrayObject} from "../../types/data";
import React, {createContext, useMemo, useState} from "react";
import {DownloadSessionContextType} from "../../types/contextTypes";
import {generateRandomString} from "../../utils/misc";
import {RequestStatus} from "../../utils/constants";

export interface SessionStatus {
    credentialTypeDisplayObj: CredentialTypeDisplayArrayObject[];
    downloadStatus: RequestStatus;
}

interface SessionsMap {
    [downloadId: string]: SessionStatus;
}

export const DownloadSessionContext = createContext<DownloadSessionContextType | undefined>(undefined);

export const DownloadSessionProvider: React.FC<{ children: React.ReactNode }> = ({children}) => {
    const [downloadInProgressSessions, setDownloadInProgressSessions] = useState<SessionsMap>({});
    const [currentSessionDownloadId, setCurrentSessionDownloadId] = useState<string | null>(null);
    const [latestDownloadedSessionId, setLatestDownloadedSessionId] = useState<string | null>(null);

    const generateUniqueDownloadId = (): string => {
        return Date.now().toString(36) + generateRandomString();
    };

    const addSession = (
        credentialTypeDisplayObj: CredentialTypeDisplayArrayObject[],
        downloadStatus: RequestStatus
    ): string => {
        const newDownloadId = generateUniqueDownloadId();

        setDownloadInProgressSessions(prevSessions => {
            const updatedSessionsMap = {
                ...prevSessions, [newDownloadId]: {credentialTypeDisplayObj, downloadStatus}
            };
            return updatedSessionsMap;
        });

        setCurrentSessionDownloadId(newDownloadId);
        return newDownloadId;
    };

    const updateSession = (downloadId: string, downloadStatus: RequestStatus) => {
        setDownloadInProgressSessions(prevSessions => {
            if (!prevSessions[downloadId]) {
                console.warn(`Attempted to update non-existent session with download ID: ${downloadId}`);
                return prevSessions;
            }

            const updatedSessionDetails = {...prevSessions[downloadId], downloadStatus};
            const updatedSessionsMap = {
                ...prevSessions,
                [downloadId]: updatedSessionDetails
            };

            return updatedSessionsMap;
        });

        if (downloadStatus === RequestStatus.DONE || downloadStatus === RequestStatus.ERROR) {
            setLatestDownloadedSessionId(downloadId)
        }
    };

    const removeSession = (downloadId: string) => {
        setDownloadInProgressSessions(prevSessions => {
            const {[downloadId]: _, ...updatedSessionsMap} = prevSessions;
            return updatedSessionsMap;
        });
        setLatestDownloadedSessionId(null);
    };

    const contextValue = useMemo(
        () => ({
            downloadInProgressSessions,
            currentSessionDownloadId,
            latestDownloadedSessionId,
            addSession,
            updateSession,
            removeSession,
            setCurrentSessionDownloadId,
            setLatestDownloadedSessionId
        }),
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [downloadInProgressSessions, currentSessionDownloadId, latestDownloadedSessionId]
    );

    return (
        <DownloadSessionContext.Provider value={contextValue}>
            {children}
        </DownloadSessionContext.Provider>
    );
};
