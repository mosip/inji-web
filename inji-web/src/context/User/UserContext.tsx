import React, {createContext, useState} from "react";
import {UserContextType} from "../../types/contextTypes";
import {ErrorType, User} from "../../types/data";
import {KEYS} from "../../utils/constants";
import {api, MethodType} from "../../utils/api";

export const UserContext = createContext<UserContextType | undefined>(undefined);

export const UserProvider: React.FC<{children: React.ReactNode}> = ({
                                                                        children
                                                                    }) => {
    const [user, setUser] = useState<User | null>(null);
    const [walletId, setWalletId] = useState<string | null>(null);
    const [error, setError] = useState<ErrorType | null>(null);
    const [isLoading, setIsLoading] = useState<boolean>(true);

    const saveUser = (userData: User) => {
        localStorage.setItem(KEYS.USER, JSON.stringify(userData));
        setUser(userData);
    };

    const removeUser = () => {
        localStorage.removeItem(KEYS.USER);
        localStorage.removeItem(KEYS.WALLET_ID);
        setUser(null);
        setWalletId(null);
    };

    const removeWallet = () => {
        localStorage.removeItem(KEYS.WALLET_ID);
        setWalletId(null);
    };

    const isUserLoggedIn = React.useMemo(() => !!user && !!walletId, [user, walletId]);

    const fetchUserProfile = async () => {
        try {
            setIsLoading(true);
            const response = await fetch(api.fetchUserProfile.url(), {
                method: MethodType[api.fetchUserProfile.methodType],
                headers: {...api.fetchUserProfile.headers()},
                credentials: api.fetchUserProfile.credentials
            });

            const responseData = await response.json();
            if (!response.ok) throw responseData;

            const userData: User = {
                displayName: responseData.displayName,
                profilePictureUrl: responseData.profilePictureUrl,
                email: responseData.email
            };

            saveUser(userData);
            setWalletId(responseData.walletId);
            localStorage.setItem(KEYS.WALLET_ID, responseData.walletId);
            setIsLoading(false);
            return {user: userData, walletId: responseData.walletId};
        } catch (error) {
            console.error('Error fetching user profile:', error);
            setError(error as ErrorType);
            removeUser();
            localStorage.removeItem(KEYS.WALLET_ID);
            setIsLoading(false);
            throw error;
        }
    };

    const contextValue = React.useMemo(
        () => ({
            user,
            walletId,
            error,
            isLoading,
            isUserLoggedIn,
            fetchUserProfile,
            saveUser,
            removeUser,
            removeWallet
        }),
        [user, walletId, error, isLoading]
    );

    return (
        <UserContext.Provider value={contextValue}>
            {children}
        </UserContext.Provider>
    );
};