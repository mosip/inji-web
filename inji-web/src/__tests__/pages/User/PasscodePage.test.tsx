import React from 'react';
import {render, screen, waitFor, within} from '@testing-library/react';
import {MemoryRouter, useNavigate} from 'react-router-dom';
import {CookiesProvider} from 'react-cookie';
import {PasscodePage} from '../../../pages/User/Passcode/PasscodePage';
import {UserProvider} from "../../../context/User/UserContext";
import {DownloadSessionProvider} from "../../../context/User/DownloadSessionContext";
import {mockApiResponse, mockApiResponseSequence, mockUseApi} from "../../../test-utils/setupUseApiMock";
import {setMockUseSelectorState} from "../../../test-utils/mockReactRedux";
import {AppStorage} from "../../../utils/AppStorage";
import userEvent from "@testing-library/user-event";
import {useUser} from "../../../hooks/User/useUser";

// Mocking the useTranslation hook from react-i18next
jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        t: (key: string) => {
            const translations: Record<string, string> = {
                "setPasscode": "Set Passcode",
                "enterPasscode": "Enter Passcode",
                "enterPasscodeLabel": "Enter Passcode",
                "confirmPasscodeLabel": "Confirm Passcode",
                "forgotPasscode": "Forgot Passcode",
                "setPasscodeDescription": "Set your passcode to get started",
                "enterPasscodeDescription": "Enter your 6 digit passcode",
                "passcodeWarning": "Make sure you remember the password for future login",
                "error.walletStatus.temporarily_locked": "You’ve reached the maximum number of attempts. Your wallet is now temporarily locked for sometime",
                "error.walletStatus.permanently_locked": "Your wallet has been permanently locked due to multiple failed attempts. Please click on forgot password to reset your wallet to continue",
                "error.walletStatus.last_attempt_before_lockout": "Incorrect passcode. Last attempt remaining before your wallet is permanently locked",
            };
            return translations[key] || key;
        }
    })
}));

jest.mock("../../../hooks/useApi.ts", () => {
    return {
        useApi: () => mockUseApi
    };
});

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
    useLocation: jest.fn()
}));

jest.mock("../../../utils/AppStorage.ts", () => ({
    AppStorage: {
        getItem: jest.fn(),
        removeItem: jest.fn(),
        setItem: jest.fn(),
    },
}))

jest.mock("../../../hooks/User/useUser.tsx", () => ({
    useUser: jest.fn(),
}))

describe('Passcode', () => {
    const mockNavigate = jest.fn();

    const renderWithProviders = (ui: React.ReactElement) => {
        return render(
            <CookiesProvider>
                <UserProvider>
                    <DownloadSessionProvider>
                        <MemoryRouter>{ui}</MemoryRouter>
                    </DownloadSessionProvider>
                </UserProvider>
            </CookiesProvider>
        );
    };

    beforeEach(() => {
        (useNavigate as jest.Mock).mockReturnValue(mockNavigate);

        setMockUseSelectorState({
            common: {
                language: 'en',
            },
        });

        (useUser as jest.Mock).mockReturnValue({
            removeWallet: jest.fn(),
            walletId: null,
            saveWalletId: jest.fn(),
            fetchUserProfile: jest.fn()
        });
    })

    test('renders passcode page', () => {
        renderWithProviders(<PasscodePage/>);
        const page = screen.getByTestId('passcode-page');
        expect(page).toBeInTheDocument();
    });

    test('renders passcode logo', () => {
        renderWithProviders(<PasscodePage/>);
        const logo = screen.getByTestId('logo-inji-web-container');
        expect(logo).toBeInTheDocument();
    });

    test('renders passcode title', () => {
        renderWithProviders(<PasscodePage/>);
        const title = screen.getByTestId('title-passcode');
        expect(title).toHaveTextContent(/Set Passcode|Enter Passcode/);
    });

    test("should render enter passcode container along with forgot passcode and submit buttons when a Wallet exists", async () => {
        mockApiResponse({
            response: [{
                "walletId": "2c2e1810-19c8-4c85-910d-aa1622412413",
                "walletName": null,
                "walletStatus": null
            }]
        });
        renderWithProviders(<PasscodePage/>);

        const passcodeInput = await screen.findByTestId("passcode-container");
        expect(passcodeInput).toBeInTheDocument();
        expect(within(passcodeInput).getByTestId("label-passcode")).toHaveTextContent("Enter Passcode");

        const forgotPasscodeButton = await screen.findByTestId("btn-forgot-passcode");
        expect(forgotPasscodeButton).toBeInTheDocument();
        expect(forgotPasscodeButton).toHaveTextContent("Forgot Passcode?");

        const submitButton = await screen.findByTestId("btn-submit-passcode");
        expect(submitButton).toBeInTheDocument();
        expect(submitButton).toBeDisabled();
    });

    test("should render enter passcode and confirm passcode containers along with submit button when Wallet doesn't exist", async () => {
        mockApiResponse({response: []});
        renderWithProviders(<PasscodePage/>);

        const passcodeInput = await screen.findByTestId("passcode-container");
        expect(passcodeInput).toBeInTheDocument();
        expect(within(passcodeInput).getByTestId("label-passcode")).toHaveTextContent("Enter Passcode");

        const confirmPasscodeInput = await screen.findByTestId("confirm-passcode-container");
        expect(confirmPasscodeInput).toBeInTheDocument();
        expect(within(confirmPasscodeInput).getByTestId("label-confirm-passcode")).toHaveTextContent("Confirm Passcode");

        const submitButton = await screen.findByTestId("btn-submit-passcode");
        expect(submitButton).toBeInTheDocument();
        expect(submitButton).toBeDisabled();
    });


    test("should redirect to forgot passcode page when forgot passcode button is clicked", async () => {
        mockApiResponse({
            response: [{
                "walletId": "2c2e1810-19c8-4c85-910d-aa1622412413",
                "walletName": null,
                "walletStatus": null
            }]
        })
        renderWithProviders(<PasscodePage/>);

        const forgotPasscodeButton = await screen.findByTestId("btn-forgot-passcode");
        expect(forgotPasscodeButton).toBeInTheDocument();
        forgotPasscodeButton.click();

        expect(mockNavigate).toBeCalledWith("/user/reset-passcode", {"state": {"walletId": "2c2e1810-19c8-4c85-910d-aa1622412413"}}
        );
    })

    test("should redirect to home when successfully unlocked wallet", async () => {
        mockApiResponseSequence([
            {
                response: [{
                    "walletId": "2c2e1810-19c8-4c85-910d-aa1622412413",
                    "walletName": null,
                    "walletStatus": null
                }]
            },
            {
                response: [{
                    "walletId": "2c2e1810-19c8-4c85-910d-aa1622412413",
                    "walletName": null,
                    "walletStatus": null
                }]
            },
        ])
        renderWithProviders(<PasscodePage/>);

        await enterPasscode()
        userEvent.click(screen.getByTestId("btn-submit-passcode"));

        await waitFor(() => expect(mockNavigate).toHaveBeenCalledTimes(1))
        expect(mockNavigate).toHaveBeenCalledWith("/user/home");
    })

    test.each([
        {
            name: "should display temporarily locked error and disable input boxes and submit button when landing on the passcode page for a temporarily locked wallet",
            walletStatus: "temporarily_locked",
            expectedError: "You’ve reached the maximum number of attempts. Your wallet is now temporarily locked for sometime",
            inputsDisabled: true,
        },
        {
            name: "should display permanently locked error and disable input boxes and submit button when landing on the passcode page for a permanently locked wallet",
            walletStatus: "permanently_locked",
            expectedError: "Your wallet has been permanently locked due to multiple failed attempts. Please click on forgot password to reset your wallet to continue",
            inputsDisabled: true,
        },
        {
            name: "should display one attempt left before lockout error and enable input boxes when landing on the passcode page for a wallet with one attempt left before permanent lockout",
            walletStatus: "last_attempt_before_lockout",
            expectedError: "Incorrect passcode. Last attempt remaining before your wallet is permanently locked",
            inputsDisabled: false,
        },
    ])(
        "$name",
        async ({walletStatus, expectedError, inputsDisabled}) => {
            mockApiResponse({
                response: [
                    {walletId: "2c2e1810-19c8-4c85-910d-aa1622412413", walletName: null, walletStatus},
                ],
            });

            renderWithProviders(<PasscodePage/>);

            verifyPasscodeInputsAndSubmitButtonState(expectedError, inputsDisabled);
        }
    );

    // TODO: Fix these test cases
    // test.each([
    //     {
    //         name: "should display temporarily locked error and disable input boxes and submit button when unlock wallet endpoint returns temporarily_locked error code",
    //         walletStatus: "temporarily_locked",
    //         expectedError: "You’ve reached the maximum number of attempts. Your wallet is now temporarily locked for sometime",
    //         inputsDisabled: true,
    //     },
    //     {
    //         name: "should display permanently locked error and disable input boxes and submit button when unlock wallet endpoint returns permanently_locked error code",
    //         walletStatus: "permanently_locked",
    //         expectedError: "Your wallet has been permanently locked due to multiple failed attempts. Please click on forgot password to reset your wallet to continue",
    //         inputsDisabled: true,
    //     },
    //     {
    //         name: "should display one attempt left before lockout error and enable input boxes when unlock wallet endpoint returns last_attempt_before_lockout error code",
    //         walletStatus: "last_attempt_before_lockout",
    //         expectedError: "Incorrect passcode. Last attempt remaining before your wallet is permanently locked",
    //         inputsDisabled: false,
    //     },
    // ])(
    //     "$name",
    //     async ({walletStatus, expectedError, inputsDisabled}) => {
    //         mockApiResponseSequence([
    //             {
    //                 response: [
    //                     {walletId: "2c2e1810-19c8-4c85-910d-aa1622412413", walletName: null, walletStatus: null},
    //                 ],
    //             },
    //             {
    //                 response: {
    //                     error: {
    //                         response: {
    //                             data: {errorCode: walletStatus},
    //                         },
    //                     },
    //                 },
    //             },
    //         ]);
    //
    //         renderWithProviders(<PasscodePage/>);
    //
    //         await enterPasscode();
    //         userEvent.click(screen.getByTestId("btn-submit-passcode"));
    //         screen.debug();
    //
    //         verifyPasscodeInputsAndSubmitButtonState(expectedError, inputsDisabled);
    //     }
    // );

    test("should enable submit button after passcode is entered for a wallet with one attempt left before permanent lockout", async () => {
        mockApiResponse({
            response: [{
                "walletId": "2c2e1810-19c8-4c85-910d-aa1622412413",
                "walletName": null,
                "walletStatus": "last_attempt_before_lockout"
            }]
        })

        renderWithProviders(<PasscodePage/>);

        const errorSpan = await screen.findByTestId("error-msg-passcode");
        expect(errorSpan).toHaveTextContent("Incorrect passcode. Last attempt remaining before your wallet is permanently locked");

        const submitButton = screen.getByTestId("btn-submit-passcode");
        expect(submitButton).toBeDisabled();

        await enterPasscode();
        expect(submitButton).not.toBeDisabled();
    })

// Testing for re-login scenario in case of session expiry
    test("should redirect to previous url post unlock if available", async () => {
        (AppStorage.getItem as jest.Mock).mockReturnValue("/previous-url");
        mockApiResponseSequence([
            {response: [{"walletId": "2c2e1810-19c8-4c85-910d-aa1622412413", "walletName": null}]},
            {response: [{"walletId": "2c2e1810-19c8-4c85-910d-aa1622412413", "walletName": null}]},
        ])
        renderWithProviders(<PasscodePage/>);

        await enterPasscode();
        userEvent.click(screen.getByTestId("btn-submit-passcode"));

        await waitFor(() =>
            expect(AppStorage.removeItem).toHaveBeenCalledWith("redirectTo", true)
        )
        expect(mockNavigate).toHaveBeenCalledTimes(1)
        expect(mockNavigate).toHaveBeenCalledWith("/previous-url");
    })

    const enterPasscode = async () => {
        await screen.findByTestId("passcode-container");
        const inputs = screen.getAllByTestId('input-passcode');
        inputs.map((input) => userEvent.type(input, '1'));
    }

    const verifyPasscodeInputsAndSubmitButtonState = async (expectedError: string, inputsDisabled: boolean) => {
        await waitFor(()=>{
            const errorSpan = screen.getByTestId("error-msg-passcode");
            expect(errorSpan).toHaveTextContent(expectedError);
        })

        const inputs = screen.getAllByTestId("input-passcode");
        inputs.forEach((input) => {
            if (inputsDisabled) {
                expect(input).toBeDisabled();
            } else {
                expect(input).not.toBeDisabled();
            }
        });

        const submitButton = screen.getByTestId("btn-submit-passcode");
        expect(submitButton).toBeDisabled();
    }
});