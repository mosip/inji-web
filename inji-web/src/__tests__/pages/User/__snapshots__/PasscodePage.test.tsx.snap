// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passcode check if layout is matching with snapshot when Wallet doesn't exist 1`] = `
<div>
  <div
    class="fixed inset-0 backdrop-blur-sm bg-black bg-opacity-40 flex flex-col items-center justify-center z-50 h-screen"
    data-testid="passcode-page"
  >
    <div
      class="rounded-2xl bg-white flex flex-col items-center justify-start relative w-[90%] sm:w-[85%] md:w-[90%] h-fit max-h-[99vh] sm:max-h-[98vh] overflow-y-hidden shadow-iw-pin-page-container"
    >
      <div
        class="overflow-hidden absolute inset-0 z-0 flex items-center justify-center pointer-events-none"
        data-testid="backdrop-passcode"
      >
        <div
          class="absolute"
          style="top: 73px;"
        >
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-0"
            style="width: 96px; height: 96px; border-width: 1px; border-color: rgba(228, 231, 236, 0.8); top: calc(50% - 48px); left: calc(50% - 48px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-1"
            style="width: 192px; height: 192px; border-width: 1px; border-color: rgba(228, 231, 236, 0.7000000000000001); top: calc(50% - 96px); left: calc(50% - 96px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-2"
            style="width: 288px; height: 288px; border-width: 1px; border-color: rgba(228, 231, 236, 0.6000000000000001); top: calc(50% - 144px); left: calc(50% - 144px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-3"
            style="width: 384px; height: 384px; border-width: 1px; border-color: rgba(228, 231, 236, 0.5); top: calc(50% - 192px); left: calc(50% - 192px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-4"
            style="width: 480px; height: 480px; border-width: 1px; border-color: rgba(228, 231, 236, 0.4); top: calc(50% - 240px); left: calc(50% - 240px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-5"
            style="width: 576px; height: 576px; border-width: 1px; border-color: rgba(228, 231, 236, 0.30000000000000004); top: calc(50% - 288px); left: calc(50% - 288px);"
          />
          <div
            class="flex items-center justify-center"
            data-testid="logo-inji-web-container"
          >
            <img
              alt="Inji Web Logo"
              data-testid="logo-inji-web"
              src="Logomark.png"
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-start w-full mt-44 sm:mt-48 relative z-10 pb-0 sm:pb-8"
      >
        <div
          class="text-center space-y-5 relative w-full max-w-lg px-4 sm:px-0"
        >
          <h1
            class="text-2xl-plus font-semibold text-iw-darkGrayishBlue"
            data-testid="title-passcode"
          >
            Set Passcode
          </h1>
          <p
            class="text-iw-textTertiary text-lg font-normal mx-8"
            data-testid="passcode-description"
          >
            Set your passcode to get started
          </p>
        </div>
        <div
          class="flex flex-col bg-white sm:rounded-lg sm:shadow-iw-pin-page-container items-center z-20 w-full max-w-lg mt-8 mb-4 mx-auto border-t border-iw-grayishBlue mx-8"
          data-testid="passcode-inputs-container"
        >
          <div
            class="w-full px-4 sm:px-8 py-3 sm:py-5 md:py-7 space-y-4 flex flex-col items-center"
          >
            <div
              class="flex flex-col gap-3.5"
            >
              <div
                class="w-full mx-auto flex flex-col gap-4"
              >
                <div
                  class="flex justify-center gap-x-2 min-w-fit"
                >
                  <div
                    class="mb-2"
                    data-testid="passcode-container"
                  >
                    <p
                      class="text-sm text-left font-medium text-iw-textSecondary"
                      data-testid="label-passcode"
                    >
                      Enter Passcode
                    </p>
                    <div
                      class="flex items-center gap-2 sm:gap-3.5"
                    >
                      <div
                        class="flex items-stretch gap-2 sm:gap-3.5 py-2 rounded-lg"
                      >
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                      </div>
                      <div
                        class="flex items-center gap-4 py-2 rounded-lg"
                      >
                        <button
                          class="pin-input-box-border pin-input-box-style flex items-center justify-center focus:outline-none"
                          data-testid="btn-toggle-visibility-passcode"
                          type="button"
                        >
                          <svg
                            class="text-iw-grayLight"
                            data-testid="eye-view-slash"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 640 512"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
                            />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="flex justify-center gap-x-2 min-w-fit"
                >
                  <div
                    class="mb-2"
                    data-testid="confirm-passcode-container"
                  >
                    <p
                      class="text-sm text-left font-medium text-iw-textSecondary"
                      data-testid="label-confirm-passcode"
                    >
                      Confirm Passcode
                    </p>
                    <div
                      class="flex items-center gap-2 sm:gap-3.5"
                    >
                      <div
                        class="flex items-stretch gap-2 sm:gap-3.5 py-2 rounded-lg"
                      >
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                        <input
                          class="pin-input-box-style pin-input-box-border focus:outline-none"
                          data-testid="input-confirm-passcode"
                          inputmode="numeric"
                          maxlength="1"
                          type="password"
                          value=""
                        />
                      </div>
                      <div
                        class="flex items-center gap-4 py-2 rounded-lg"
                      >
                        <button
                          class="pin-input-box-border pin-input-box-style flex items-center justify-center focus:outline-none"
                          data-testid="btn-toggle-visibility-confirm-passcode"
                          type="button"
                        >
                          <svg
                            class="text-iw-grayLight"
                            data-testid="eye-view-slash"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 640 512"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
                            />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="w-full"
              >
                <div
                  class="w-full"
                >
                  <button
                    class="w-full bg-iw-primary text-center items-center text-iw-text font-bold text-sm px-6 py-3 rounded-lg shadow text-white mr-1 false bg-gray-400 cursor-not-allowed grayscale"
                    data-testid="btn-submit-passcode"
                    disabled=""
                  >
                    <div
                      class="flex items-center justify-center gap-1.5"
                    >
                      <span>
                        submit
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Passcode check if layout is matching with snapshot when Wallet exists 1`] = `
<div>
  <div
    class="fixed inset-0 backdrop-blur-sm bg-black bg-opacity-40 flex flex-col items-center justify-center z-50 h-screen"
    data-testid="passcode-page"
  >
    <div
      class="rounded-2xl bg-white flex flex-col items-center justify-start relative w-[90%] sm:w-[85%] md:w-[90%] h-fit max-h-[99vh] sm:max-h-[98vh] overflow-y-hidden shadow-iw-pin-page-container"
    >
      <div
        class="overflow-hidden absolute inset-0 z-0 flex items-center justify-center pointer-events-none"
        data-testid="backdrop-passcode"
      >
        <div
          class="absolute"
          style="top: 73px;"
        >
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-0"
            style="width: 96px; height: 96px; border-width: 1px; border-color: rgba(228, 231, 236, 0.8); top: calc(50% - 48px); left: calc(50% - 48px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-1"
            style="width: 192px; height: 192px; border-width: 1px; border-color: rgba(228, 231, 236, 0.7000000000000001); top: calc(50% - 96px); left: calc(50% - 96px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-2"
            style="width: 288px; height: 288px; border-width: 1px; border-color: rgba(228, 231, 236, 0.6000000000000001); top: calc(50% - 144px); left: calc(50% - 144px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-3"
            style="width: 384px; height: 384px; border-width: 1px; border-color: rgba(228, 231, 236, 0.5); top: calc(50% - 192px); left: calc(50% - 192px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-4"
            style="width: 480px; height: 480px; border-width: 1px; border-color: rgba(228, 231, 236, 0.4); top: calc(50% - 240px); left: calc(50% - 240px);"
          />
          <div
            class="absolute rounded-full border overflow-hidden"
            data-testid="backdrop-passcode-circle-5"
            style="width: 576px; height: 576px; border-width: 1px; border-color: rgba(228, 231, 236, 0.30000000000000004); top: calc(50% - 288px); left: calc(50% - 288px);"
          />
          <div
            class="flex items-center justify-center"
            data-testid="logo-inji-web-container"
          >
            <img
              alt="Inji Web Logo"
              data-testid="logo-inji-web"
              src="Logomark.png"
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-start w-full mt-44 sm:mt-48 relative z-10 pb-0 sm:pb-8"
      >
        <div
          class="text-center space-y-5 relative w-full max-w-lg px-4 sm:px-0"
        >
          <h1
            class="text-2xl-plus font-semibold text-iw-darkGrayishBlue"
            data-testid="title-passcode"
          >
            Enter Passcode
          </h1>
          <p
            class="text-iw-textTertiary text-lg font-normal mx-8"
            data-testid="passcode-description"
          >
            Enter your 6 digit passcode
          </p>
        </div>
        <div
          class="flex flex-col bg-white sm:rounded-lg sm:shadow-iw-pin-page-container items-center z-20 w-full max-w-lg mt-8 mb-4 mx-auto border-t border-iw-grayishBlue mx-8"
          data-testid="passcode-inputs-container"
        >
          <div
            class="w-full px-4 sm:px-8 py-3 sm:py-5 md:py-7 space-y-4 flex flex-col items-center"
          >
            <div
              class="flex flex-col gap-3.5"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
